### PR TITLE
document config option "url_fetch_method" 

### PR DESCRIPTION
**Pull request by loulawrence**
See https://github.com/spack/spack/pull/21398 and https://github.com/spack/spack/pull/24622

- Change `config: use_curl: True | False`  to `config: url_fetch_method: curl | urllib` for clarity 
- Documentation of url_fetch_method in defaults/config.yaml
- Default fetch option explicitly set to `urllib` for users who may not have curl on their system

### CODE REVIEW

**Review by adamjstewart — DISMISSED** [review]
Thanks for documenting this, this looks much better! A lot of users have requested `wget` support because their `curl` installation is broken, so I think `url_fetch_method` is much more generic and extensible than `use_curl`.

We may still want to add better error messages or a way to detect whether Python's urllib installation actually works, but that can be saved for another PR.

**Review by adamjstewart — DISMISSED** [review]

**Comment by haampie** [comment]
Any reason why this wasn't merged yet?

**Comment by haampie** [comment]
@spackbot re-run pipeline

**Comment by adamjstewart** [comment]
> Any reason why this wasn't merged yet?

Nope, just waiting for a second review. Go ahead and review/merge.

**Comment by haampie** [comment]
@spackbot re-run pipeline

File: etc/spack/defaults/config.yaml
Comment (haampie) [inline etc/spack/defaults/config.yaml:140]: Should we just call it

```suggestion
  fetch_method: urllib
```

?

(Sorry for starting a bike shedding discussion...)

Comment (adamjstewart) [reply etc/spack/defaults/config.yaml:140]: There are other non-URL fetch methods, such as git/svn/hg.

Comment (haampie) [reply etc/spack/defaults/config.yaml:140]: But they also use urls `https://github.com/spack/spack.git` :thinking:

Comment (haampie) [reply etc/spack/defaults/config.yaml:140]: Ok, let's keep it `url_fetch_method`, since curl calls itself the tool "for transferring data with URLs" and urllib is called urllib.

**Review by haampie — CHANGES_REQUESTED** [review]
Can you add `url_fetch_method` to the schema and add upgrade logic? Basically this:

```diff
diff --git a/lib/spack/spack/schema/config.py b/lib/spack/spack/schema/config.py
index 1df922b137..0cf533eb18 100644
--- a/lib/spack/spack/schema/config.py
+++ b/lib/spack/spack/schema/config.py
@@ -99,6 +99,10 @@
             },
             'allow_sgid': {'type': 'boolean'},
             'binary_index_root': {'type': 'string'},
+            'url_fetch_method': {
+                'type': 'string',
+                'enum': ['urllib', 'curl']
+            },
         },
     },
 }
@@ -153,4 +157,10 @@ def update(data):
         update_data = spack.config.merge_yaml(update_data, projections_data)
         data['install_tree'] = update_data
         changed = True
+
+    use_curl = data.pop('use_curl', None)
+    if use_curl is not None:
+        data['url_fetch_method'] = 'curl' if use_curl else 'urllib'
+        changed = True
+
     return changed
```

~Otherwise looks good to me.~ Actually we don't have this option in the docs, do we? That's important to have too...

**Comment by BetsyMcPhail** [comment]
> ~Otherwise looks good to me.~ Actually we don't have this option in the docs, do we? That's important to have too...

Can you point me to where the docs should be updated?

**Review by haampie — APPROVED** [review]
Actually it's fine, the docs will reflect it automatically because they include the config.yaml file.

**Pull request merged**
